client/tso_stream.go

```diff
@@ -172,51 +178,255 @@ func (s tsoTSOStreamAdapter) Recv() (tsoRequestResult, error) {
+	prevState := s.state.Swap(streamStateSending)
+	switch prevState {
+	case streamStateIdle:
+		// Expected case
+	case streamStateClosing:
+		s.state.Store(prevState)
+		err := s.GetRecvError()
+		log.Info("sending to closed tsoStream", zap.String("stream", s.streamID), zap.Error(err))
+		if err == nil {
+			err = errors.WithStack(errs.ErrClientTSOStreamClosed)
 		}
-		return
+		return err
+	case streamStateSending:
+		log.Fatal("unexpected concurrent sending on tsoStream", zap.String("stream", s.streamID))
+	default:
+		log.Fatal("unknown tsoStream state", zap.String("stream", s.streamID), zap.Int32("state", prevState))
+	}
+
+	select {
+	case s.pendingRequests <- batchedRequests{
+		startTime:          start,
+		count:              count,
+		reqKeyspaceGroupID: keyspaceGroupID,
+		callback:           callback,
+	}:
+	default:
+		s.state.Store(prevState)
+		return errors.New("unexpected channel full")
+	}
+	s.state.Store(prevState)
+
+	if err := s.stream.Send(clusterID, keyspaceID, keyspaceGroupID, dcLocation, count); err != nil {
+		// As the request is already put into `pendingRequests`, the request should finally be canceled by the recvLoop.
+		// So skip returning error here to avoid
+		// if err == io.EOF {
+		// 	return errors.WithStack(errs.ErrClientTSOStreamClosed)
+		// }
+		// return errors.WithStack(err)
+		log.Warn("failed to send RPC request through tsoStream", zap.String("stream", s.streamID), zap.Error(err))
+		return nil
 	}
 	tsoBatchSendLatency.Observe(time.Since(batchStartTime).Seconds())
-	res, err := s.stream.Recv()
-	duration := time.Since(start).Seconds()
-	if err != nil {
-		requestFailedDurationTSO.Observe(duration)
-		if err == io.EOF {
-			err = errs.ErrClientTSOStreamClosed
-		} else {
-			err = errors.WithStack(err)
+	s.ongoingRequestCountGauge.Set(float64(s.ongoingRequests.Add(1)))
+	return nil
+}
+
+func (s *tsoStream) recvLoop(ctx context.Context) {
+	var finishWithErr error
+	var currentReq batchedRequests
+	var hasReq bool
+
+	defer func() {
+		if r := recover(); r != nil {
+			log.Fatal("tsoStream.recvLoop internal panic", zap.Stack("stacktrace"), zap.Any("panicMessage", r))
 		}
-		return
+
+		if finishWithErr == nil {
+			// The loop must exit with a non-nil error (including io.EOF and context.Canceled). This should be
+			// unreachable code.
+			log.Fatal("tsoStream.recvLoop exited without error info", zap.String("stream", s.streamID))
+		}
+
+		if hasReq {
+			// There's an unfinished request, cancel it, otherwise it will be blocked forever.
+			currentReq.callback(tsoRequestResult{}, currentReq.reqKeyspaceGroupID, finishWithErr)
+		}
+
+		s.stoppedWithErr.Store(&finishWithErr)
+		s.cancel()
+		for !s.state.CompareAndSwap(streamStateIdle, streamStateClosing) {
+			switch state := s.state.Load(); state {
+			case streamStateIdle, streamStateSending:
+				// streamStateSending should switch to streamStateIdle very quickly. Spin until successfully setting to
```

Comment:
Spin？

Reply:
Yes. Considering that there won't be any blocking operation in `streamStateSending` state and the state won't last long, I think spinning here is better than writing another syncing operation here.